metagraph/src/cli/load/load_graph.cpp

```diff
@@ -71,5 +71,47 @@ std::shared_ptr<DeBruijnGraph> load_critical_dbg(const std::string &filename) {
     exit(1);
 }
 
+template <class Graph>
+std::shared_ptr<CanonicalDBG> primary_to_canonical(std::shared_ptr<Graph> graph,
+                                                   size_t cache_size) {
+    if (graph->get_mode() != DeBruijnGraph::PRIMARY) {
+        logger->error("Only primary mode graphs can be wrapped into canonical mode.");
+        exit(1);
+    }
+
+    logger->trace("Primary graph wrapped into canonical");
+    return std::make_shared<CanonicalDBG>(graph, cache_size);
+}
+
+template std::shared_ptr<CanonicalDBG> primary_to_canonical(std::shared_ptr<const DeBruijnGraph>, size_t);
+template std::shared_ptr<CanonicalDBG> primary_to_canonical(std::shared_ptr<DeBruijnGraph>, size_t);
+
+std::shared_ptr<const DeBruijnGraph> make_cached_graph(const DeBruijnGraph &graph,
+                                                       const Config &config,
+                                                       size_t cache_size) {
+    std::shared_ptr<const DeBruijnGraph> ret_graph(std::shared_ptr<const DeBruijnGraph>{}, &graph);
```

Comment:
Currently, you can map wrap first to canonical with cache and then into a cached view, or the other way around.
Isn't one of these more optimal than the other? Did you think about this?
I think we should allow only a certain order of wrapping, to ensure things are wrapped optimally.

Reply:
UP: Since canonical wrapper needs some caches, maybe it just makes sense to keep a cached view on the graph in it, and use its cache?

Then, if you want to create another cached view on that CanonicalDBG, you would copy the internal cache and create another CanonicalDBG<CachedView>.
So in addition to the initial CanonicalDBG<CachedView> you'll create another CanonicalDBG<CachedView> with independent cache, initialized from the first one, so you don't lose any things computed beforehand.

Reply:
The optimal wrapping strategy depends on the task. If you're only interested in exact k-mer query against the graph and annotations, then CanonicalDBG alone is enough for its mapping functions and won't use any caches. So, having the underlying graph be uncached is optimal.

However, for alignment, it's more optimal to have the underlying graph be a cached view since it can cache decoded nodes from map_to_nodes calls.

Reply:
Also, as mentioned below, having a cache leads to `map_to_nodes` locking for every node so it can populate the cache, so it's less optimal for use as a global object (i.e., each thread should have its own cached view).